rust/crd/src/resource.rs

```diff
@@ -31,132 +78,262 @@ impl RoleResource {
-#[allow(clippy::enum_variant_names)]
-pub enum Error {
-    #[snafu(display("no resources available for merging"))]
-    NoResourcesToMerge,
-    #[snafu(display("cannot merge storage types of different roles"))]
-    IncompatibleStorageMerging,
+lazy_static! {
+    pub static ref DEFAULT_RESOURCES: ResourcesFragment<storage::DruidStorage, NoRuntimeLimits> =
+        ResourcesFragment {
+            cpu: CpuLimitsFragment {
+                min: Some(Quantity("200m".to_owned())),
+                max: Some(Quantity("4".to_owned())),
+            },
+            memory: MemoryLimitsFragment {
+                limit: Some(Quantity("2Gi".to_owned())),
+                runtime_limits: NoRuntimeLimitsFragment {},
+            },
+            storage: storage::DruidStorageFragment {},
+        };
+    pub static ref HISTORICAL_RESOURCES: ResourcesFragment<storage::HistoricalStorage, NoRuntimeLimits> =
+        ResourcesFragment {
+            cpu: CpuLimitsFragment {
+                min: Some(Quantity("200m".to_owned())),
+                max: Some(Quantity("4".to_owned())),
+            },
+            memory: MemoryLimitsFragment {
+                limit: Some(Quantity("2Gi".to_owned())),
+                runtime_limits: NoRuntimeLimitsFragment {},
+            },
+            storage: storage::HistoricalStorageFragment {},
+        };
+}
+
+fn default_resources(role: &DruidRole) -> Option<RoleResourceFragment> {
+    match role {
+        DruidRole::Historical => Some(RoleResourceFragment::HistoricalFragment(
+            HISTORICAL_RESOURCES.clone(),
+        )),
+        _ => Some(RoleResourceFragment::DruidFragment(
+            DEFAULT_RESOURCES.clone(),
+        )),
+    }
+}
+
+fn role_resources(druid: &DruidCluster, role: &DruidRole) -> Option<RoleResourceFragment> {
+    match role {
+        DruidRole::Broker => druid
+            .spec
+            .brokers
+            .clone()
+            .config
+            .config
+            .resources
+            .map(RoleResourceFragment::DruidFragment),
+        DruidRole::Coordinator => druid
+            .spec
+            .coordinators
+            .clone()
+            .config
+            .config
+            .resources
+            .map(RoleResourceFragment::DruidFragment),
+        DruidRole::Historical => druid
+            .spec
+            .historicals
+            .clone()
+            .config
+            .config
+            .resources
+            .map(RoleResourceFragment::HistoricalFragment),
+        DruidRole::MiddleManager => druid
+            .spec
+            .middle_managers
+            .clone()
+            .config
+            .config
+            .resources
+            .map(RoleResourceFragment::DruidFragment),
+        DruidRole::Router => druid
+            .spec
+            .routers
+            .clone()
+            .config
+            .config
+            .resources
+            .map(RoleResourceFragment::DruidFragment),
+    }
```

Comment:
This would clone less. Can we get around working with refs completely?
```suggestion
    match role {
        DruidRole::Broker => druid
            .spec
            .brokers
            .config
            .config
            .resources
            .clone()
            .map(RoleResourceFragment::DruidFragment),
        DruidRole::Coordinator => druid
            .spec
            .coordinators
            .config
            .config
            .resources
            .clone()
            .map(RoleResourceFragment::DruidFragment),
        DruidRole::Historical => druid
            .spec
            .historicals
            .config
            .config
            .resources
            .clone()
            .map(RoleResourceFragment::HistoricalFragment),
        DruidRole::MiddleManager => druid
            .spec
            .middle_managers
            .config
            .config
            .resources
            .clone()
            .map(RoleResourceFragment::DruidFragment),
        DruidRole::Router => druid
            .spec
            .routers
            .config
            .config
            .resources
            .clone()
            .map(RoleResourceFragment::DruidFragment),
    }
```